HPCC-17644 Initial work on remote key delivery

jakesmith opened the pull request:
<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [ ] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [ ] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [ ] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browers
  - [ ] The component(s) render as expected

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

HPCCSmoketest commented:
Automated Smoketest: :x:
Sha: 1dcdbb088e9ad277fc87ffbcc09ae378bb953325
Build: failed
Number of Error(s): 104
/mnt/disk1/home/vamosax/smoketest/PR-10016/HPCC-Platform/system/jlib/jbuff.hpp:287:1: error: expected primary-expression before \xe2\x80\x98<<\xe2\x80\x99 token
/mnt/disk1/home/vamosax/smoketest/PR-10016/HPCC-Platform/system/jlib/jbuff.hpp:287:3: error: expected primary-expression before \xe2\x80\x98<<\xe2\x80\x99 token
/mnt/disk1/home/vamosax/smoketest/PR-10016/HPCC-Platform/system/jlib/jbuff.hpp:287:5: error: expected primary-expression before \xe2\x80\x98<<\xe2\x80\x99 token
/mnt/disk1/home/vamosax/smoketest/PR-10016/HPCC-Platform/system/jlib/jbuff.hpp:287:7: error: expected primary-expression before \xe2\x80\x98<\xe2\x80\x99 token
/mnt/disk1/home/vamosax/smoketest/PR-10016/HPCC-Platform/system/jlib/jbuff.hpp:287:9: error: \xe2\x80\x98HEAD\xe2\x80\x99 was not declared in this scope
/mnt/disk1/home/vamosax/smoketest/PR-10016/HPCC-Platform/system/jlib/jbuff.hpp:288:9: error: expected \xe2\x80\x98;\xe2\x80\x99 before \xe2\x80\x98mb\xe2\x80\x99
/mnt/disk1/home/vamosax/smoketest/PR-10016/HPCC-Platform/system/jlib/jbuff.hpp:289:1: error: expected primary-expression before \xe2\x80\x98==\xe2\x80\x99 token
/mnt/disk1/home/vamosax/smoketest/PR-10016/HPCC-Platform/system/jlib/jbuff.hpp:289:3: error: expected primary-expression before \xe2\x80\x98==\xe2\x80\x99 token
/mnt/disk1/home/vamosax/smoketest/PR-10016/HPCC-Platform/system/jlib/jbuff.hpp:289:5: error: expected primary-expression before \xe2\x80\x98==\xe2\x80\x99 token
/mnt/disk1/home/vamosax/smoketest/PR-10016/HPCC-Platform/system/jlib/jbuff.hpp:289:7: error: expected primary-expression before \xe2\x80\x98=\xe2\x80\x99 token
/m
 ... ( comment is too long, 111 lines, 17127 bytes, truncated) ... 
... make[2]: *** [system/jlib/CMakeFiles/jlib.dir/jbuff.cpp.o] Error 1
make[2]: *** [system/jlib/CMakeFiles/jlib.dir/jcomp.cpp.o] Error 1
make[1]: *** [system/jlib/CMakeFiles/jlib.dir/all] Error 2
make: *** [all] Error 2
Get ECLWatch build result


```

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 9297da88991ceb2d565299f3a7e435091f9ec271
Build: success
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 93.868 seconds |


Install hpccsystems-platform-community_6.3.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |93 |93 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |10 |10 |0 |
|setup (thor) |10 |10 |0 |
|setup (roxie) |10 |10 |0 |
|test (hthor) |728 |728 |0 |
|test (thor) |622 |621 |1 |
|test (roxie) |749 |749 |0 |


Errors:
- test
  - thor
    - steplimit1 (multiPart=true), 

HPCC Stop: OK
HPCC Uninstall: OK

jakesmith commented:
@richardkchapman - ready to review, but needs further testing

richardkchapman commented on thorlcr/activities/indexread/thindexreadslave.cpp:
```diff
@@ -246,20 +235,24 @@ class CIndexReadSlaveBase : public CSlaveActivity
+            Owned<IKeyManager> klManager;
 
-//            if (!rfn.isLocal() && !localKey && !seekGEOffset)
+            //if (!rfn.isLocal() && !localKey && !seekGEOffset)
```
Is this commented out for testing?

jakesmith replied:
It's been replaced in a newer commit, where it's also dependent upon a #option.

richardkchapman commented on thorlcr/graph/thgraphslave.cpp:
```diff
@@ -1775,14 +1775,34 @@ IReplicatedFile *createEnsurePrimaryPartFile(CActivityBase &activity, const char
 class CFileCache;
 class CLazyFileIO : public CInterface, implements IFileIO, implements IDelayedFile
 {
+    class CFileIOWrapper : public CSimpleInterfaceOf<IFileIO>
```
Is this change required for this issue or is it something else you spotted needing fixing?

jakesmith replied:
Should be removed, I will do so.
Though there's a slight caveat/oddity involved here that I'm not sure how to address.
The remote key manager gets the underlying 'real' IFileIO from the lazy io via getRealFileIO().
Which means it can live slightly longer and outside the lazy fileIO cache.

The lazyfileio cache updates the read stats on removal (on IFileIO::close).
Since the remote io doesn't involve any direct IFileIO calls, that's okay as far as these stats are concerned.
Local key stats would be kept via jhtree 
Remote key stats would be tracked remotely, but should probably be captured here too somehow.

richardkchapman commented on common/remote/remoteerr.hpp:
```diff
@@ -66,6 +66,7 @@
 #define RFSERR_VersionMismatch                  8047
 #define RFSERR_SetThrottleFailed                8048
 #define RFSERR_MaxQueueRequests                 8049
+#define RFSERR_KedIndexFailed                   8050
```
Is this supposed to say KeyIndexFailed?

jakesmith replied:
yep, I'll change.

richardkchapman commented on common/remote/remoteerr.hpp:
```diff
@@ -86,6 +87,7 @@
 #define RAERR_InvalidUsernamePassword_Text      "Invalid (upper case U) in username/password"
```
Is this still an error??

jakesmith replied:
I dunno, no change here, apart from spurious new blank line below it (which I'll remove)

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 9297da88991ceb2d565299f3a7e435091f9ec271
Build: success
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 93.209 seconds |


Install hpccsystems-platform-community_6.3.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |93 |93 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |10 |10 |0 |
|setup (thor) |10 |10 |0 |
|setup (roxie) |10 |10 |0 |
|test (hthor) |728 |728 |0 |
|test (thor) |622 |621 |1 |
|test (roxie) |749 |749 |0 |


Errors:
- test
  - thor
    - steplimit1 (multiPart=false), 

HPCC Stop: OK
HPCC Uninstall: OK

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 80235d66ea83dbd189524c3972bea00ca3fd2768
Build: success
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 93.82 seconds |


Install hpccsystems-platform-community_6.3.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |93 |93 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |10 |10 |0 |
|setup (thor) |10 |10 |0 |
|setup (roxie) |10 |10 |0 |
|test (hthor) |728 |728 |0 |
|test (thor) |622 |621 |1 |
|test (roxie) |749 |749 |0 |


Errors:
- test
  - thor
    - steplimit1 (multiPart=false), 

HPCC Stop: OK
HPCC Uninstall: OK

richardkchapman commented on dali/dafilesrv/dafilesrv.cpp:
```diff
@@ -691,6 +692,8 @@ int main(int argc,char **argv)
     startPerformanceMonitor(10*60*1000, PerfMonStandard, &perfHook);
     writeSentinelFile(sentinelFile);
+
+    setKeyIndexCacheSize(0); // disable IKeyIndex caching
```
Why?

jakesmith replied:
Need to discuss.
If enabled, or if not cleared - key writes will not be reflected, i.e. remote key reads will pick up old copies.

ghalliday replied:
We don't generally overwrite files once written, what is the likely affect of disabling the cache?

jakesmith replied:
I think we need to revisit this separately.
In new cleaned up PR (about to open) I have a slightly better solution I think, where instead of disabling, I keep the cache and flush the specific key on client close.
That at least avoids the problem you do hit otherwise where a updated key (e.g. format) hits errors.

I do wonder if disabling the cache completely might be best in dafilesrv and therefore instead rely on the disk cache.
Unfortunately setKeyIndexCacheSize(0) does not really disable it completely, the jhtree key code is so heavily inter-twined with the node cache that it will still use it.
It is in need of some refactoring as we have discussed before for other reasons too.

jakesmith committed:
HPCC-15675 Refactor lazy fileio

1) To avoid critical section being held too long - it was being held for
duration of i/o operation, now it will be held whilst the IFileIO
interface is fetched and linked only.
2) Avoid a bug whereby a file had been added to the cache by an old
activity then opened by a new and hit a failure, in doing so it would
reference the old (released) activity and crash. Refactor so the current
activity requesting the cached item passes in the activity.

Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

ghalliday commented on common/remote/sockfile.cpp:
```diff
@@ -2582,6 +2593,172 @@ void CRemoteFile::copyTo(IFile *dest, size32_t buffersize, ICopyFileProgress *pr
+        segs.append(segment);
+    }
+    virtual unsigned ordinality() const override { UNIMPLEMENTED; }
+    virtual IKeySegmentMonitor *item(unsigned idx) const override { UNIMPLEMENTED; }
```
segs.item(i)?

jakesmith replied:
yes

ghalliday commented on common/remote/sockfile.cpp:
```diff
@@ -3160,6 +3338,15 @@ class CClientStatsTable : public OwningStringSuperHashTableOf<CClientStats>
     }
 };
 
+struct OpenFileInfo
```
minor/future: Code is fine as it is.  Chould possibly have this as a a linked structure, filename a StringAttr, and have a local instance of this which is filled in and then appended to a CArrayOf<> if the open succeeded.  As it is there will be 32 bytes being copied per file instead of 8 when 1 is removed.

Should possibly be a hash table or linked list( instead of an array since there will probably be quite a few entries, and elements get added and removed out of order.  (A std::map which uses a red-black tree may give the best of both worlds.)

jakesmith replied:
Yes, room for improvement here. I thought should be hash table, but didn't want to change at this juncture.

ghalliday commented on common/remote/sockfile.cpp:
```diff
@@ -3830,6 +4022,78 @@ class CRemoteFileServer : implements IRemoteFileServer, public CInterface
+        CriticalBlock block(sect);
+        unsigned clientidx;
+        unsigned handleidx;
+        if (!findHandle(handle,clientidx,handleidx))
```
Out of interest, what is the average number of open files on a dafilesrv?

jakesmith replied:
I don't know, but one of the reasons I thought (surprised wasn't / had forgotten) that it was using a hash table for lookup..

ghalliday commented on common/remote/sockfile.cpp:
```diff
@@ -3830,6 +4022,78 @@ class CRemoteFileServer : implements IRemoteFileServer, public CInterface
+        unsigned handleidx;
+        if (!findHandle(handle,clientidx,handleidx))
+        {
+            VStringBuffer msg("Failed to find existing key manager for contiuation. key = %s", keyname);
```
typo: conti*n*uation

jakesmith replied:
code has changed in this area, now uses serialized key cursor.

ghalliday commented on common/remote/sockfile.cpp:
```diff
@@ -3830,6 +4022,78 @@ class CRemoteFileServer : implements IRemoteFileServer, public CInterface
+        IKeyManager *kM = client.openFiles.item(handleidx).keyManager;
+        if (nullptr == kM)
+        {
+            VStringBuffer msg("Failed to find existing key manager for contiuation. key = %s", keyname);
```
typo: conti*n*uation

jakesmith replied:
as above

ghalliday reviewed (commented):
@jakesmith see comments from a first review

HPCCSmoketest commented:
Automated Smoketest: :x:
Sha: 48bb4f9835a6858046716f12391ef583f8d63608
Build: failed
Number of Error(s): 10
/mnt/disk1/home/vamosax/smoketest/PR-10016/HPCC-Platform/system/jlib/jbuff.hpp:252:1: error: expected unqualified-id before \xe2\x80\x98<<\xe2\x80\x99 token
/usr/include/c++/4.8.2/typeinfo:39:37: error: expected declaration before end of line
make[2]: *** [system/jlib/CMakeFiles/jlib.dir/jargv.cpp.o] Error 1
/mnt/disk1/home/vamosax/smoketest/PR-10016/HPCC-Platform/system/jlib/jbuff.hpp:252:1: error: expected unqualified-id before \xe2\x80\x98<<\xe2\x80\x99 token
/usr/include/c++/4.8.2/typeinfo:39:37: error: expected declaration before end of line
make[2]: *** [system/jlib/CMakeFiles/jlib.dir/jarray.cpp.o] Error 1
make[1]: *** [system/jlib/CMakeFiles/jlib.dir/all] Error 2
[  0%] Building CXX object aws-cpp-sdk-core/CMakeFiles/aws-cpp-sdk-core.dir/source/client/CoreErrors.cpp.o
[100%] Building CXX object aws-cpp-sdk-sqs/CMakeFiles/aws-cpp-sdk-sqs.dir/source/SQSErrors.cpp.o
make: *** [all] Error 2
Get ECLWatch build result

jakesmith committed:
HPCC-17644 Rationalize dafilesrv IFileIO/handle/filename arrays into struct

Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

jakesmith committed:
HPCC-17644 Fix CSetKeySegmentMonitor deserialization bug.

lastCompareValue was not being correctly initialized.

Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

jakesmith committed:
HPCC-17644 Add IKeyManager utility method setSegmentMonitors

To make it easier manage a deserialized SegMonitorList

Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

jakesmith committed:
HPCC-17644 Support for remote key filtering + counting/limits.

Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

jakesmith committed:
HPCC-17644 Add forceDafilesrv testing option for indexes

Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: dad805c508bcdd1b549e9f1859666f80775faa15
Build: success
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 94.059 seconds |


Install hpccsystems-platform-community_6.3.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |94 |94 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |730 |730 |0 |
|test (thor) |624 |624 |0 |
|test (roxie) |758 |757 |1 |


Errors:
- test
  - roxie
    - soapcall (targetIP=127001, goodPort=9876, blacListedPort=9875), 

HPCC Stop: OK
HPCC Uninstall: OK

jakesmith committed:
HPCC-17644 Flush key store cache on key close.

Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

jakesmith commented on common/remote/sockfile.cpp:
```diff
@@ -2582,6 +2593,172 @@ void CRemoteFile::copyTo(IFile *dest, size32_t buffersize, ICopyFileProgress *pr
+        maxRecsPerRequest = 100;
+        totalGot = 0;
+    }
+    virtual void releaseSegmentMonitors() override { segs.segMonitors.kill(); }
```
Agree. I think I added SegMonitorList::reset() later and forgot to change this.

jakesmith commented on common/remote/sockfile.cpp:
```diff
@@ -2582,6 +2593,172 @@ void CRemoteFile::copyTo(IFile *dest, size32_t buffersize, ICopyFileProgress *pr
+                {
+                    if (totalGot >= chooseNLimit)
+                        break;
+                    maxRecs = chooseNLimit - totalGot;
```
agree, will change

jakesmith commented on common/remote/sockfile.cpp:
```diff
@@ -2582,6 +2593,172 @@ void CRemoteFile::copyTo(IFile *dest, size32_t buffersize, ICopyFileProgress *pr
+    {
+        segs.append(segment);
+    }
+    virtual unsigned ordinality() const override { UNIMPLEMENTED; }
```
yes could be, although not foreseen to be used yet. A couple of other seg methods could be implemented too.

jakesmith committed:
HPCC-17644 A few changes following GH review.

Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

jakesmith commented:
Closing in favour of cleaner PR.
Review changes will be most recent commit.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 406e842e5e6942ef04595bedf6eaefb587bdad93
Build: success
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 94.058 seconds |


Install hpccsystems-platform-community_6.3.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |94 |94 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |730 |730 |0 |
|test (thor) |624 |624 |0 |
|test (roxie) |758 |758 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
